### ISSUE TITLE
[Bug]: ReferenceError: process is not defined

### ISSUE THREAD

**Issue by dhanushkumarsivaji** [web-infra-dev/rspack#5955]
### System Info

```
  "devDependencies": {
    "@rspack/cli": "0.5.4",
    "@rspack/core": "0.5.4",
    "@rspack/plugin-react-refresh": "0.5.4",
    "@types/react": "^18.2.0",
    "@types/react-dom": "^18.2.0",
    "autoprefixer": "^10.1.0",
    "postcss": "^8.2.1",
    "postcss-loader": "^8.0.0",
    "react-refresh": "^0.14.0"
  },
  "dependencies": {
    "@rsbuild/core": "^0.4.15",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  }
```

### Details

rspack.config.js

```
const rspack = require('@rspack/core')
const refreshPlugin = require('@rspack/plugin-react-refresh')
const isDev = process.env.NODE_ENV === 'development'


/**
 * @type {import('@rspack/cli').Configuration}
 */
module.exports = {
  context: __dirname,
  entry: {
    main: './src/index.js',
  },
  devServer: {
    historyApiFallback: true,
    port: 3000,
  },
  resolve: {
    extensions: ['.js','.jsx','.ts','.tsx','.json']
  },
  module: {
    rules: [
      {
        test: /\.svg$/,
        type: 'asset',
      },
      {
        test: /\.scss$/,
        use: [
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: {
                  tailwindcss: {},
                  autoprefixer: {},
                },
              },
            },
          },
        ],
        type: 'css',
      },
      {
        test: /\.(jsx?|tsx?)$/,
        use: [
          {
            loader: 'builtin:swc-loader',
            options: {
              sourceMap: true,
              jsc: {
                parser: {
                  syntax: 'typescript',
                  tsx: true,
                },
                transform: {
                  react: {
                    runtime: 'automatic',
                    development: isDev,
                    refresh: isDev,
                  },
                },
              },
              env: {
                targets: [
                  'chrome >= 87',
                  'edge >= 88',
                  'firefox >= 78',
                  'safari >= 14',
                ],
              },
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new rspack.container.ModuleFederationPlugin({
      name: 'host',
      filename: 'remoteEntry.js',
      exposes: {},
      shared: {
        react: { eager: true },
        'react-dom': { eager: true },
        'react-router-dom': { eager: true },
      },
    }),
    new rspack.DefinePlugin({
      'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
    }),
    new rspack.ProgressPlugin({}),
    new rspack.HtmlRspackPlugin({
      template: './src/index.html',
    }),
    isDev ? new refreshPlugin() : null,
  ].filter(Boolean),
}

```
.env

```
APP_NAME="DEV"

```
App.js

```
import React from "react";
import ReactDOM from "react-dom";


const App = () => (
  <>
    {process.env.APP_NAME}
  </>
);
ReactDOM.render(<App />, document.getElementById("app"));

```

Also I have different environment files like, .env.development, .env.uat-wu, .env.uat-eu, .env.prod-wu, .env.prod-eu

I want to load this different environement files based on the environment. I have update the script like this below

```
  "scripts": {
    "build": "NODE_ENV=production rspack build",
    "build:dev": "NODE_ENV=development rspack build",
    "build:uat-eu": "NODE_ENV=uat-eu rspack build",
    "build:uat-wu": "NODE_ENV=uat-wu rspack build",
    "build:start": "cd dist && rspack serve",
    "start": "NODE_ENV=development rspack serve",
    "start:uat-eu": "NODE_ENV=uat-eu rspack serve",
    "start:uat-wu": "NODE_ENV=uat-wu rspack serve"
  }

```

Also please let me know how to use this different environment files.

I am right now in migrating from webpack to rspack. But faced this problem and also need to know how to use this environment files.

below is my webpack file, I need to convert it into rspack

```
const Dotenv = require('dotenv-webpack');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');
const TerserPlugin = require('terser-webpack-plugin');
const ModuleFederationPlugin = require('webpack/lib/container/ModuleFederationPlugin');
const ExternalTemplateRemotesPlugin = require('external-remotes-plugin');
const deps = require('../package.json').dependencies;
const env = process.env.NODE_ENV;

module.exports = {
  mode: 'development',
  output: {
    publicPath: 'http://localhost:3000/',
  },
  resolve: {
    extensions: ['.js', '.jsx'],
  },
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
          options: {
            presets: ['@babel/preset-react', '@babel/preset-env'],
            plugins: ['@babel/plugin-transform-runtime'],
          },
        },
      },
      {
        test: /\.(sass|css|scss)$/,
        use: ['style-loader', 'css-loader', 'postcss-loader', 'sass-loader'],
      },
      {
        test: /.(png|jpeg|gif|svg)$/,
        type: 'asset/resource',
      },
      {
        test: /.(ttf|woff|woff2)$/,
        type: 'asset/resource',
      },
    ],
  },
  devServer: {
    port: 3000,
    historyApiFallback: true,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
      'Access-Control-Allow-Headers': 'X-Requested-With, content-type, Authorization',
    },
  },
  plugins: [
    new ModuleFederationPlugin({
      name: 'container',
      filename: 'remoteEntry.js',
      remotes: {
        // companyRanking: 'companyRanking@http://localhost:3002/remoteEntry.js',
        holdings: 'holdings@http://localhost:3003/remoteEntry.js',
      },
      exposes: {
        './useTokenManagementStore': './src/store/tokenManagement',
      },
      shared: {
        ...deps,
        react: { singleton: true, requiredVersion: deps.react },
        'react-dom': { singleton: true, requiredVersion: deps['react-dom'] },
        'ews-ui-library': { singleton: true, requiredVersion: deps['ews-ui-library'] },
      },
    }),
    new ExternalTemplateRemotesPlugin(),
    new HtmlWebpackPlugin({
      template: './public/index.html',
      favicon: './public/favicon.ico',
      manifest: './public/manifest.json',
    }),
    new Dotenv({
      path: `./.env.${env}`,
    }),
  ],
  optimization: {
    minimizer: [
      new CssMinimizerPlugin(),
      new TerserPlugin({
        terserOptions: {
          format: {
            comments: false,
          },
        },
        extractComments: false,
      }),
    ],
  },
};

```

Would be really appreciated will I be getting help in converting the above config file to rspack

Thanks in advance
    
   

### Reproduce link

https://github.com/dhanushkumarsivaji/rspack-module-federation

### Reproduce Steps

npm start

**Comment by xc2** [comment]
as before, you need `dotenv-webpack` to make the dotenv files work.

however, unfortunately, the dotenv-webpack will not work with rspack before https://github.com/mrsteele/dotenv-webpack/pull/501 is merged.

In the meantime, you can manually copy the content of this file as a plugin: https://github.com/mrsteele/dotenv-webpack/blob/f32653e03ac75ef087527992cc27f52a7d412efc/src/index.js

Alternatively, you may consider migrating from webpack to [rsbuild](https://rsbuild.dev). Rsbuild is an out-of-box builder built upon rspack and comes with a built-in [dotenv loading feature](https://rsbuild.dev/guide/advanced/env-vars#default-variables). May you enjoy it.

**Comment by xc2** [comment]
I'm closing this issue as the dotenv-webpack has released a new version 8.1.0 with rspack support, which will meet what you wanna achieve. please give a try.
